### ISSUE TITLE
Dragging a tab can show visual glitch when tab not fully visible

### ISSUE THREAD

**Issue by ericdrobinson** [Microsoft/vscode#38450]
# Description
An image is worth a thousand words:
![](https://i.imgur.com/zZVxR6a.gif)

But an image is not good for discoverability...

Moving a tab that is partially hidden/obscured by the debug widget will cause the tab to draw incorrectly: the portion of the tab visually blocked by the widget will show that part of the widget (it's 100% drawn over it).

# Version Info:
- VSCode Version: 1.18.0
- OS Version: macOS 10.12.6

# Steps to Reproduce:

1. Attach the Debugger to a process.
2. Move the Debug Actions Pane such that it partially obscures a tab.
3. Try moving the partially obscured tab to another position.

# Errata
**Reproduces without extensions:** Didn't get it to work - needed the Chrome Debugger Extension.

**Comment by isidorn** [comment]
Duplicate of https://github.com/Microsoft/vscode/issues/2513

**Comment by ericdrobinson** [comment]
@isidorn I beg to differ:
#2513 is about the debug panel being *in front of tabs.*
This bug is about the *rendering of tabs when moved* when they happen to be behind the debug panel.

Please do not conflate the two. Fixing #2513 would not necessarily fix this bug _*unless*_ you assume that the fix for that bug would move the panel.

**Comment by weinand** [comment]
Yes, this is not a duplicate of #2513 but a bug that we'll have to fix.

**Comment by bpasero** [comment]
The only way how we can fix this is by providing our own custom HTML element in the drop feedback, otherwise this is just chromes standard behaviour when you DND.

**Comment by bpasero** [comment]
https://github.com/Microsoft/vscode/issues/18733

**Comment by weinand** [comment]
@bpasero Would it be possible to hide the debug action floating window before starting the drag?

**Comment by bpasero** [comment]
@weinand not sure that is a good idea, after all there can be many more things laying on top of the tabs (for example a message or quick open) that would cause this issue. I think a better approach would be to somehow lift the tabs on top of anything else that could be on top once you drag. Haven't tried that though.

**Comment by weinand** [comment]
Yes, bringing the tabs to front might be the simpler approach.
But is it easily possible to restore the old z-ordering at the end of the drag?

**Comment by bpasero** [comment]
@weinand yes, but its not easy to just make the tab position `absolute` and `zindex` because that is not the style of the tab in the first place. We are using flex layout.

**Comment by ericdrobinson** [comment]
@weinand @bpasero Would it be possible to make use of the [DataTransfer.setDragImage](https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/setDragImage) API here? According to [this site](http://mereskin.github.io/dnd/) it can take arbitrary HTML elements as input _as well as_ images.

**Comment by bpasero** [comment]
@ericdrobinson yeah if you wanna give it a try, feel free.

**Comment by bpasero** [comment]
@ericdrobinson hold on, I found another trick that might do it without having to use the `setDragImage` API.

**Comment by bpasero** [comment]
Fixed via https://github.com/Microsoft/vscode/commit/e1eeccd55cf0bbbe378efd8e04745714eefb41c2

**Comment by weinand** [comment]
BTW, great trick!
